acceptancetests/assess_upgrade.py

```diff
@@ -0,0 +1,199 @@
+            client.juju('list-users', (), include_e=False)
+            return True
+        except CalledProcessError:
+            # Upgrade will still be in progress and thus refuse the api call.
+            return False
+    try:
+        wait_for_model_check(client, model_upgrade_status_check, timeout)
+    except ModelCheckFailed:
+        raise JujuAssertionError(
+            'Upgrade for model {} failed to complete within the alloted '
+            'timeout ({} seconds)'.format(
+                client.model_name, timeout))
+
+
+def assert_stable_model_is_correct(stable_client):
+    assert_model_is_version(
+        stable_client.get_controller_client(),
+        get_stripped_version_number(stable_client.version))
+    assert_model_is_version(
+        stable_client,
+        get_stripped_version_number(stable_client.version))
+
+
+def assert_model_is_version(client, expected_version, timeout=600):
+    # Check agent & machine version
+    client.wait_for_version(expected_version, timeout)
+    # Also check model version
+    client.wait_for(WaitModelVersion(expected_version, timeout))
+
+
+def setup_agent_metadata(stream_server, agent, client, tmp_dir, stream):
+    version_parts = get_version_parts(client.version)
+
+    if agent is None:
+        agent_details = agent_tgz_from_juju_binary(client.full_path, tmp_dir)
+    else:
+        agent_details = agent
+
+    stream_server.add_product(
+        stream,
+        version_parts.version,
+        version_parts.arch,
+        client.env.get_option('default-series'),
+        agent_details)
+    # Trusty needed for wikimedia charm.
+    stream_server.add_product(
+        stream,
+        version_parts.version,
+        version_parts.arch,
+        'trusty',
+        agent_details)
+
+
+def get_version_parts(version_string):
+    parts = version_string.split('-')
+    return VersionParts(parts[0], parts[1], parts[2])
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser(
+        description='Test juju upgrades for controllers and models')
+    add_basic_testing_arguments(parser, existing=False)
+    parser.add_argument(
+        '--stable-juju-bin',
+        help='Path to juju binary to be used as the stable version of juju.')
+    parser.add_argument(
+        '--stable-agent',
+        help='Path to agent to use when bootstrapping with stable binary'
+    )
+    parser.add_argument(
+        '--devel-agent',
+        help='Path to agent to use when bootstrapping with stable binary'
+    )
+    return parser.parse_args(argv)
+
+
+def main(argv=None):
+    args = parse_args(argv)
+    configure_logging(args.verbose)
+
+    stable_bsm = get_stable_juju(args, args.stable_juju_bin)
+    devel_client = stable_bsm.client.clone(
+        version=ModelClient.get_version(args.juju_bin),
+        full_path=args.juju_bin,
+        debug=args.debug
+    )
+
+    assess_upgrade_from_stable_to_develop(args, stable_bsm, devel_client)
```

Comment:
since we are intending others to add-on scenarios to test, ot MAY be useful in this specific case to add a high verbosity comment here. Want to make sure adding secondary scenarios is easy / feels good. So things like environments, etc.